### PR TITLE
Update subscription configuration schema to include new parameters

### PR DESCRIPTION
**Pull request by benbp**
This change simplifies cloud configurations in our live test pipelines. Currently these parameters get defined per cloud in the live test template, which ends up creating a large, verbose configuration. To simplify these definitions, I'd like to move them into the subscription configuration. The resulting template usage becomes more like:

```
- name: CloudConfig
  type: object
  default:
    Public:
      SubscriptionConfiguration: $(sub-config-azure-cloud-test-resources)
    Preview:
      SubscriptionConfiguration: $(sub-config-azure-cloud-test-resources-preview)
    Canary:
      SubscriptionConfiguration: $(sub-config-azure-cloud-test-resources-preview)
      Location: 'eastus2euap'                                                                                                                                                                                                                                                                UsGov:
      SubscriptionConfiguration: $(sub-config-gov-test-resources)
    China:
      SubscriptionConfiguration: $(sub-config-cn-test-resources)
```

This change will require an update to the above 4 subscription configurations. The script change handles both the old and new data structure.

Passing live test referencing the new configuration: https://dev.azure.com/azure-sdk/internal/_build/results?buildId=600922&view=logs&j=bbd7d436-e05d-51a8-99a8-0701dc9dffeb&t=20f065cd-bf8b-5b9d-5626-dd2113fcfcf1

Passing live test referencing the old configuration: https://dev.azure.com/azure-sdk/internal/_build/results?buildId=600932&view=results

### CODE REVIEW

File: eng/common/TestResources/deploy-test-resources.yml [hunk 1]
```diff
@@ -27,13 +42,31 @@ steps:
         ${{ parameters.SubscriptionConfiguration }}
       "@ | ConvertFrom-Json -AsHashtable;
 
+      if ($subscriptionConfiguration.ArmTemplateParameters) {
+        # Handle subscription configuration containing arm template parameters
+        $armTemplateParameters = $subscriptionConfiguration.ArmTemplateParameters
```
Comment (weshaggard) [inline eng/common/TestResources/deploy-test-resources.yml:47]: Instead of blindly cutting over couldn't we set the ArmTemplateParameters yml parameter's default value to be `$subscriptionConfiguration.ArmTemplateParameters` and then we go and remove any places that pass a value and test it as you remove them? 

Right now as it stands this will ignore any parameters.ArmTemplateParameters everywhere as soon as you put it in KV and so there will be dead parameters and we wouldn't have even tested the change before we changed this under everyone.

File: eng/common/TestResources/deploy-test-resources.yml [hunk 2]
```diff
@@ -27,13 +42,31 @@ steps:
...
+      } else {
+        # Handle legacy subscription configuration, with arm template params passed in by templates
+        $armTemplateParameters = ${{ parameters.ArmTemplateParameters }}
+      }
+
+      $accessParameters = @{
```
Comment (weshaggard) [inline eng/common/TestResources/deploy-test-resources.yml:53]: What is the point of copying this over? Is it to remove ArmTemplateParameters? If so I think it would be better to just go ahead and pass those as well. I believe they will be ignored for now but we could eventually stop explicitly passing AdditionalParameters and start consuming them from here.

File: eng/common/TestResources/deploy-test-resources.yml (hunk 1 above)
Comment (benbp) [reply eng/common/TestResources/deploy-test-resources.yml:47]: Yep, good catch, the default logic should have picked the local string. I can change it to default parameters.

File: eng/common/TestResources/deploy-test-resources.yml (hunk 2 above)
Comment (benbp) [reply eng/common/TestResources/deploy-test-resources.yml:53]: We chatted offline, but for context: the New-TestResources.ps1 treats the new fields as unknown parameters when they are splatted in as arguments. We decided to add all the parameters as bindings to the script, since eventually we'll want to handle all of them there.

File: eng/common/TestResources/deploy-test-resources.yml [hunk 3]
```diff
@@ -27,13 +42,31 @@ steps:
         ${{ parameters.SubscriptionConfiguration }}
       "@ | ConvertFrom-Json -AsHashtable;
 
+      if ($subscriptionConfiguration.ArmTemplateParameters) {
```
Comment (heaths) [inline eng/common/TestResources/deploy-test-resources.yml:45]: While you're ad it, I would appreciate if you didn't restrict specifying additional parameters to only CIs. My intent was always for tests.yml to be able to provide them as well. Since you're merging hashtables, now is a great opportunity to provide for that.

**Comment by benbp** [comment]
@heaths @weshaggard  the latest update includes a couple of things, based on our conversations.

1. Support the old syntax used in a few places (mainly storage tests for java and js) that involved passing a powershell hash literal down as a parameter string to be rendered directly. This way I can phase this out independently.
2. Support subscription configuration defining the base set of arm template parameters, if (1.) is not defined.
3. Add pipeline and/or platform specific arm parameters, passing yaml objects down as pipeline parameters, to be converted. These will all get merged with the arm template parameters from the subscription configuration and passed down to the New-TestResources script as one object.

File: eng/common/TestResources/deploy-test-resources.yml [hunk 4]
```diff
@@ -42,6 +43,7 @@ steps:
         ${{ parameters.SubscriptionConfiguration }}
       "@ | ConvertFrom-Json -AsHashtable;
 
+<<<<<<< HEAD
```
Comment (weshaggard) [inline eng/common/TestResources/deploy-test-resources.yml:46]: looks like a unfinished merge.

Comment (benbp) [reply eng/common/TestResources/deploy-test-resources.yml]: Fixed. Sorry about that. Testing the changes in other repos and didn't inspect my own copy 👎

File: eng/common/TestResources/New-TestResources.ps1 [hunk 5]
```diff
@@ -59,7 +59,10 @@ param (
     [string] $Environment = 'AzureCloud',
 
     [Parameter()]
-    [hashtable] $AdditionalParameters,
+    [hashtable] $ArmTemplateParameters,
```
Comment (heaths) [inline eng/common/TestResources/New-TestResources.ps1:62]: This is a breaking change. This script isn't just used by CIs, but by devs. Why this change?

Comment (heaths) [reply eng/common/TestResources/New-TestResources.ps1:62]: The script already removes variables that aren't declared as parameters from the ARM template.

File: eng/common/TestResources/deploy-test-resources.yml [hunk 6]
```diff
@@ -15,24 +17,62 @@ parameters:
...
+      $subscriptionConfiguration = @'
         ${{ parameters.SubscriptionConfiguration }}
-      "@ | ConvertFrom-Json -AsHashtable;
+      '@ | ConvertFrom-Json -AsHashtable;
+
+      # Support legacy ArmTemplateParameters passed in directly as a hashmap, otherwise default to subscriptionConfiguration JSON.
+      #   Example legacy input: @{ 'param1' = 'foo' ; 'param2' = 'bar' }
+      $subscriptionConfiguration.ArmTemplateParameters = ${{ parameters.ArmTemplateParameters }} ? ${{ parameters.ArmTemplateParameters }} : @{}
+
+      # Support additional sdk pipeline and/or platform specific arm template parameters:
```
Comment (heaths) [inline eng/common/TestResources/deploy-test-resources.yml:51]: In this example, probably good to show the property name as well so there's no confusion:
```yml
ArmTemplateParameters:
  EnableHsm:
    # ...
```

**Review by heaths — CHANGES_REQUESTED** [review]
There's no need to split parameters and environment variables - a breaking change. Parameters are already filtered to what the ARM template declares.

File: eng/common/TestResources/New-TestResources.ps1 (hunk 5 above)
Comment (weshaggard) [reply eng/common/TestResources/New-TestResources.ps1:62]: Agreed we should add a new parameter instead of replacing it. We can join them later in the script as needed.

File: eng/common/TestResources/deploy-test-resources.yml [hunk 7]
```diff
@@ -1,9 +1,11 @@
 parameters:
   ServiceDirectory: not-set
-  ArmTemplateParameters: '@{}'
+  ArmTemplateParameters: '$subscriptionConfiguration.ArmTemplateParameters'
```
Comment (weshaggard) [inline eng/common/TestResources/deploy-test-resources.yml:3]: If we do end up having AdditionalParameters and ArmTemplateParameters you can keep the empty list as the default.

File: eng/common/TestResources/deploy-test-resources.yml [hunk 8]
```diff
@@ -15,24 +17,62 @@ parameters:
...
+      # Support additional sdk pipeline and/or platform specific arm template parameters:
+      #   EnableHsm:
+      #     Value: true
+      #     Platforms:
+      #       - Linux_NetCore_Keyring
+      #   TestGlobalParam:
+      #     Value: global foo
+      # For duplicate keys, prefer the template/script parameter.
+      $platformArmTemplateParameters = @'
+        ${{ parameters.PlatformArmTemplateParameters }}
```
Comment (weshaggard) [inline eng/common/TestResources/deploy-test-resources.yml:60]: Should we continue to use AdditionalParameters here and add combine those inside the script?

File: eng/common/TestResources/New-TestResources.ps1 (hunk 5 above)
Comment (heaths) [reply eng/common/TestResources/New-TestResources.ps1:62]: Do we even need to do that, though? The script already passes all parameters to the test-resources-post.ps1 script (if present), so why separate variables? At the very least, any parameters passed in should be merged to pass on through.

Comment (weshaggard) [reply eng/common/TestResources/New-TestResources.ps1:62]: I like moving a lot of the default parameters into the sub-config and I'm definitely open for allowing some tests to enable passing additional parameters in special cases. It is just a matter of where we merge them. I suggested merging the script so that we don't have to do a lot of inline scripting outside the script and we don't have to do much remapping of the sub-config to special case this.

Comment (heaths) [reply eng/common/TestResources/New-TestResources.ps1:62]: I don't think we're talking about the same thing. I'm saying the `AdditionalParameters` (i.e. not changing it to `ArmTemplateParameters`) and `EnvironmentVariables` don't need to be separate. Extraneous variables of any kind won't be passed to the ARM template. So why not just pass them all in as `AdditionalParameters`?

Comment (benbp) [reply eng/common/TestResources/New-TestResources.ps1:62]: I didn't realize this script was used by devs in other places, hence I thought I could control the breaking change. Thanks for pointing that out :) Can either of you provide more context on who/how people use this? I think the logic sharing is good, but at the same time having devs take a dependency on our provisioning scripts means we can basically never make breaking changes without some sort of non-standardized effort to notify people. I'm not sure this is a good situation to be in.

Re: AdditionalParameters, I think the name could be more descriptive. The parameters are only used for ARM templates, but the naming does not make that clear. Similarly, the type is different as AdditionalParameters/ArmTemplateParameters is an object. Someone looking at a tests.yml file may be confused trying to understand the difference between matrix variables (string type) defined in AdditionalPlatforms items, and parameters defined in a field named AdditionalParameters. If they are only intended to be used for arm templates, I think we should call them that.

The other reason we wanted to add an ArmTemplateParameters argument, was because the planned update to the subscription configuration added a field called the same. Adding the argument enables us to still be able to splat the subscription configuration object. It seems to me like for the short term, we should rename that field to AdditionalParameters and keep the parameter here as AdditionalParameters just so we don't have all these merging layers of multiple variables that all serve the same purpose. But I'd like to know how easy/hard it is to ascertain the dev impact of making a breaking change to this script (...maybe some telemetry would be nice, ha).

Comment (heaths) [reply eng/common/TestResources/New-TestResources.ps1:62]: "AdditionalParameters" is a common term in PowerShell modules, hence why I chose it. They weren't necessarily meant for any particular context. While the focus of the scripts was to deploy an ARM template, pre- and post-scripts can get those as well and do whatever a service needs. Since I strip the undefined parameters out when passed to ARM (deployment fails otherwise), these really were meant to be just extra parameters for whatever is needed.

The scripts were written so devs and the CI have a single, unified way of deploying resources. We use them a lot. For example, when I write new tests I need to record the sessions, so I run the script, run/record my tests (fix them or the code if needed), commit my changes, and remove the resource group. It's the same script and mostly the same processing for devs and CIs this way. Before, every repo and most services within those repos was solving this differently, so everyone had to figure it out. Clean-up also wasn't prevalent. The scripts unify the approach now.

Comment (benbp) [reply eng/common/TestResources/New-TestResources.ps1:62]: That's helpful context, thanks!

So aside from naming, the main constraint with needing a new parameter is wanting to switch pipelines away from having to define stringified powershell objects for additional parameters, and instead being able to use yaml objects instead. In order to this with backwards compat for various tests.yml files (which I plan to migrate), we need a separate parameter with an object type.

One limitation I didn't catch earlier is that azure pipelines cannot render yaml objects into strings. However they are planning to [add support for this](https://developercommunity.visualstudio.com/idea/880210/allow-type-casting-or-expression-function-from-yam.html). As a result, I'm thinking of editing this script a little to [build the object in the powershell script](https://github.com/microsoft/azure-pipelines-yaml/issues/427) string based on the new format. Then once we can to yaml->string conversions, both the parameter definition and backing code will be simplified.

Comment (weshaggard) [reply eng/common/TestResources/New-TestResources.ps1:62]: Ok I did make an assumption that "AdditionalParameters" was only for the ARM template parameters. I still like the idea of having explicit new parameters for specific scenarios like env variables and arm template parameters. If needed we can merge in AdditionalParameters in with the ArmTemplateParameters then filter before we deploy but overtime it would be good for folks to pass template parameters under the name ArmTemplateParameters so that it is clear what they are intended for.

Comment (benbp) [reply eng/common/TestResources/New-TestResources.ps1:62]: Actually, that approach might not be possible either, in which case we'll need to just use powershell object literals as strings for the moment :(

Comment (heaths) [reply eng/common/TestResources/New-TestResources.ps1:62]: PowerShell doesn't require stringified parameters. I use these all the time like so:
```powershell
New-TestResources.ps1 ... -AdditionalParameters @{enableHsm=$true}
```
Works fine. It's YAML that's requiring this, hence peoplehaving to pass `"true"` instead of `$true`.

Originally AdditionalParameters only goes to ARM template, yes. It wasn't necessarily meant to. Adding more parameters makes this harder to use - not easier. If no one else is using AdditionalParameters for anything other than ARM (easy to search quick), just use it for ARM parameters but keep the name.

My point was, though, I don't see why you need to split env vars and ARM template parameters. Why can't both be in the same hashtable?

Comment (benbp) [reply eng/common/TestResources/New-TestResources.ps1:62]: It's not powershell that's the problem, it's passing objects around to be rendered into string blocks that azure pipelines doesn't support. Pipelines cannot convert yaml to a string, hence why the existing templates pass around a stringified hash literal.

As far as I could tell across all repos it's only being used as arm template parameters, and even then only in three places for storage/cosmos (two in java, one in js).

Comment (benbp) [reply eng/common/TestResources/New-TestResources.ps1:62]: The way I see the issue now is we have to use a stringified object literal, which will either be a yaml string or a powershell object string (I favor the latter at the moment for consistency). The decisions to make are whether this gets defined as a platform variable or a pipeline variable (I favor the latter), and whether to introduce a new variable or not.

Comment (benbp) [reply eng/common/TestResources/New-TestResources.ps1:62]: Env vars also get defined at the template parameter level, not the platform variable level. The pipelines using custom arm template parameters today differ on a per-platform basis, so we can't merge the two.

Comment (heaths) [reply eng/common/TestResources/New-TestResources.ps1:62]: Per discussion, let's add `[hashtable] $ArmTemplateParameters` and `[hashtable] $EnvironmentVariables` separately. Docs should probably clearly state these should be hashtables of string, string

Be sure to merge `$ArmTemplateParameters` into `$AdditionalParameters` so the types therein aren't lost (e.g. `$true`).

File: eng/common/TestResources/New-TestResources.ps1 [hunk 9]
```diff
@@ -313,11 +319,14 @@ if ($TenantId) {
 if ($TestApplicationSecret) {
     $templateParameters.Add('testApplicationSecret', $TestApplicationSecret)
 }
+if ($ArmTemplateParameters) {
+    $templateParameters += $ArmTemplateParameters
+}
 if ($AdditionalParameters) {
     $templateParameters += $AdditionalParameters
```
Comment (heaths) [inline eng/common/TestResources/New-TestResources.ps1:326]: Not that we're merging two unknowns, this will fail if there's duplicates. You might want to loop over `AdditionalParameters` and overwrite instead:
```powershell
foreach ($key in $AdditionalParameters.Keys) { # No need to check for null; properties in PowerShell will never throw.
  $templateParameters[$key] = $AdditionalParameters[$Key]
}
```

**Comment by benbp** [comment]
@heaths @weshaggard updated based on our meeting. For context: due to the object->string rendering limitations with devops yaml, and breaking change concerns with dev usage of the script, I'm adding ArmTemplateParameters as a net new parameter, that's defined as a stringified powershell hash in pipeline yaml like we do in other repos today.

I'm also setting all the new environment variables defined in the new subscription configurations as deployment outputs. Looking at these values more closely, I'm wondering if they are better off being defined in the sdk pipelines instead. Another discussion though.

File: eng/common/TestResources/New-TestResources.ps1 [hunk 10]
```diff
@@ -379,6 +388,10 @@ foreach ($templateFile in $templateFiles) {
         "$($serviceDirectoryPrefix)STORAGE_ENDPOINT_SUFFIX" = $context.Environment.StorageEndpointSuffix;
     }
 
+    foreach ($ev in $EnvironmentVariables) {
+        $deploymentOutputs[$ev.Name] = $ev.Value
```
Comment (heaths) [inline eng/common/TestResources/New-TestResources.ps1:392]: I worry a bit this could overwrite variables atop. Are we reasonably certain that wouldn't happen? Otherwise, should we maybe call `$deploymentOutputs.Add($ev.Name, $ev.Value)` here to make sure it throws (alerting us of a problem) instead?

**Review by heaths — APPROVED** [review]
Remember to regenerate the markdown as well. IIRC, the README has instructions how.

File: eng/common/TestResources/New-TestResources.ps1 (hunk 5 above)
Comment (benbp) [reply eng/common/TestResources/New-TestResources.ps1:62]: Updated. Since we're passing raw powershell text, we aren't constrained by string:string anymore.

File: eng/common/TestResources/New-TestResources.ps1 (hunk 9 above)
Comment (benbp) [reply eng/common/TestResources/New-TestResources.ps1:326]: Ah, good point. In my head I didn't register the operator behavior for hashes.

File: eng/common/TestResources/New-TestResources.ps1 (hunk 10 above)
Comment (heaths) [reply eng/common/TestResources/New-TestResources.ps1:392]: ...and worried only about these variables given they are both calculated outputs (well, some are inputs, but credential-related ones). Overwriting those could be disastrous, but if that's unlikely this is perfectly fine.

Comment (benbp) [reply eng/common/TestResources/New-TestResources.ps1:392]: It just depends which we think has higher priority. I was thinking we'd want to overwrite any values, as if they are the subscription config I'd assume they are known overrides necessary for a cloud to work. I'm fine with either logging or throwing though just to make sure this behavior isn't hidden.

**Review by heaths — APPROVED** [review]

File: eng/common/TestResources/deploy-test-resources.yml [hunk 11]
```diff
@@ -15,24 +15,47 @@ parameters:
...
+      $subscriptionConfiguration = @'
         ${{ parameters.SubscriptionConfiguration }}
-      "@ | ConvertFrom-Json -AsHashtable;
+      '@ | ConvertFrom-Json -AsHashtable;
+
+      if (!$subscriptionConfiguration.ArmTemplateParameters) {
+        $subscriptionConfiguration.ArmTemplateParameters = @{}
+      }
+
+      $armTemplateParameters = ${{ parameters.ArmTemplateParameters }}
```
Comment (heaths) [inline eng/common/TestResources/deploy-test-resources.yml:49]: Doesn't this need to be declared a parameter? And in the parent template as well so I can pass it through from sdk/keyvault/tests.yml? Or is declaring parameters not *required*? Even still, it makes them more discoverable.

Comment (benbp) [reply eng/common/TestResources/deploy-test-resources.yml:49]: ArmTemplateParameters is declared as a parameter at the top of this file. Or are you referring to something else?

It will get passed through like so, with ArmTemplateParameters coming from a variable declared in AdditionalPlatforms.

```
      - template: /eng/common/TestResources/deploy-test-resources.yml
        parameters:
          Location: ${{ parameters.CloudConfig.Location }}
          ServiceDirectory: '${{ parameters.ServiceDirectory }}'
          SubscriptionConfiguration: ${{ parameters.CloudConfig.SubscriptionConfiguration }}
          ArmTemplateParameters: $(ArmTemplateParameters)
```

So you would need to declare ArmTemplateParameters on your platform definition in tests.yml, and we pass it through from the matrix entry. I haven't pushed the PR in azure-sdk-for-net yet since I'm still working out an issue with when the var is empty.

File: eng/common/TestResources/New-TestResources.ps1 [hunk 12]
```diff
@@ -65,7 +65,8 @@ param (
     [hashtable] $AdditionalParameters,
 
     [Parameter()]
-    [hashtable] $EnvironmentVariables,
+    [ValidateNotNullOrEmpty()]
```
Comment (heaths) [inline eng/common/TestResources/New-TestResources.ps1:68]: This works (won't require the parameter, just that it contains data), but you might want to allow empty. Pointless to pass it, but more permissions:
```suggestion
    [ValidateNotNull()]
```

**Review by heaths — APPROVED** [review]

File: eng/common/TestResources/New-TestResources.ps1 (hunk 12 above)
Comment (benbp) [reply eng/common/TestResources/New-TestResources.ps1:68]: Based on how powershell works, if we pass it in as empty, will it then not get initialized by the default value?

File: eng/common/TestResources/deploy-test-resources.yml [hunk 13]
```diff
@@ -42,21 +42,16 @@ steps:
...
       eng/common/TestResources/New-TestResources.ps1 `
         -BaseName 'Generated' `
         -ServiceDirectory '${{ parameters.ServiceDirectory }}' `
         -Location '${{ parameters.Location }}' `
         -DeleteAfterHours '${{ parameters.DeleteAfterHours }}' `
+        # The subscriptionConfiguration may have ArmTemplateParameters defined, so
+        # pass those in via the ArmTemplateParameters flag, and handle any
+        # additional parameters from the pipelines via AdditionalParameters
         @subscriptionConfiguration `
+        -AdditionalParameters ${{ parameters.ArmTemplateParameters }} `
```
Comment (heaths) [inline eng/common/TestResources/deploy-test-resources.yml:54]: You said "pass those via the ArmTemplateParameters flag" (technically "parameter"), but then you pass to `AdditionalParameters`. Either the comment doesn't line up with the change, or I'm just confused.

Comment (benbp) [reply eng/common/TestResources/deploy-test-resources.yml:54]: It's confusing. subscriptionConfiguration has an ArmTemplateParameters key, so we set that param (which is a new param as of this PR) when we splat @subscriptionConfiguration. But we still want to pass in any ones defined by the pipeline, and merge them all together in the script. So it leverages AdditionalParameters as a way to do that.

Comment (benbp) [reply eng/common/TestResources/deploy-test-resources.yml:54]: It would be slightly easier to understand what's going on if we merged it here, but we're trying to avoid putting any more logic than necessary in the pipeline templates.

Comment (heaths) [reply eng/common/TestResources/deploy-test-resources.yml:54]: Ah, okay, yes. I see it now. Thanks for clarifying!

File: eng/common/TestResources/New-TestResources.ps1 [hunk 14]
```diff
@@ -322,11 +328,14 @@ if ($TenantId) {
 if ($TestApplicationSecret) {
     $templateParameters.Add('testApplicationSecret', $TestApplicationSecret)
 }
-if ($AdditionalParameters) {
-    $templateParameters += $AdditionalParameters
+if ($ArmTemplateParameters) {
```
Comment (weshaggard) [inline eng/common/TestResources/New-TestResources.ps1:331]: We should also look and set the variables here just in case there are any parameters being overridden.  I would also like us to add some warning messages if we are overridding a value because that will most likely cause issues and it would be nice to be able to diagnose that if it happens.

File: eng/common/TestResources/New-TestResources.ps1 [hunk 15]
```diff
@@ -388,6 +397,10 @@ foreach ($templateFile in $templateFiles) {
         "$($serviceDirectoryPrefix)STORAGE_ENDPOINT_SUFFIX" = $context.Environment.StorageEndpointSuffix;
     }
 
+    foreach ($key in $EnvironmentVariables.Keys) {
+        $deploymentOutputs.Add($key, $EnvironmentVariables[$key])
```
Comment (weshaggard) [inline eng/common/TestResources/New-TestResources.ps1:401]: This will throw an exception if there are conflicts. Should we instead just output a warning if there is a conflict? That might make diagnosing a little easier.

File: eng/common/TestResources/deploy-test-resources.yml [hunk 16]
```diff
@@ -15,25 +15,43 @@ parameters:
...
 
+
 steps:
   - template: /eng/common/TestResources/setup-az-modules.yml
 
   - pwsh: |
       eng/common/TestResources/Import-AzModules.ps1
 
-      $subscriptionConfiguration = @"
+      $subscriptionConfiguration = @'
```
Comment (weshaggard) [inline eng/common/TestResources/deploy-test-resources.yml:41]: Where we hitting some cases where PS was translating variables? I'm curious why the change from `"` to `'`?

File: eng/common/TestResources/New-TestResources.ps1 (hunk 15 above)
Comment (heaths) [reply eng/common/TestResources/New-TestResources.ps1:401]: Originally it my my intent when I suggested it to throw - so we know there was an issue. But emitting a warning may be just as good. Hopefully nothing anybody does with this could be destructive. My fear was that some parameter gets overridden that could be very bad, but hopefully that it just limited to new resources and not existing resources...like someone deleting the wrong Key Vault accidentally.

File: eng/common/TestResources/deploy-test-resources.yml (hunk 16 above)
Comment (heaths) [reply eng/common/TestResources/deploy-test-resources.yml:41]: Best practice. It's always safer if you don't intend variables to be expanded. I recommend using single quotes in PowerShell all the time if you don't need double quotes. Even if it's not malicious, it can certainly mess things up.

File: eng/common/TestResources/New-TestResources.ps1 [hunk 17]
```diff
@@ -101,6 +108,16 @@ function Retry([scriptblock] $Action, [int] $Attempts = 5) {
     }
 }
 
+function MergeHashes([hashtable] $source, [psvariable] $dest) {
+    foreach ($key in $source.Keys) {
+        if ($dest.Value[$key]) {
+            Write-Warning ("Overwriting '$($dest.Name).$($key)' with value '$($dest.Value[$key])' " +
```
Comment (weshaggard) [inline eng/common/TestResources/New-TestResources.ps1:114]: We should use our devops logging functions here. It doesn't have to happen in this PR but at some point we should switch to using those.

File: eng/common/TestResources/New-TestResources.ps1 [hunk 18]
```diff
@@ -289,7 +305,7 @@ if ($CI) {
     # Set the resource group name variable.
     Write-Host "Setting variable 'AZURE_RESOURCEGROUP_NAME': $ResourceGroupName"
     Write-Host "##vso[task.setvariable variable=AZURE_RESOURCEGROUP_NAME;]$ResourceGroupName"
-    $environmentVariables['AZURE_RESOURCEGROUP_NAME'] = $ResourceGroupName
+    $EnvironmentVariables['AZURE_RESOURCEGROUP_NAME'] = $ResourceGroupName
```
Comment (weshaggard) [inline eng/common/TestResources/New-TestResources.ps1:308]: This could also potentially overwrite although not likely.

File: eng/common/TestResources/New-TestResources.ps1 [hunk 19]
```diff
@@ -101,6 +108,16 @@ function Retry([scriptblock] $Action, [int] $Attempts = 5) {
     }
 }
 
+function MergeHashes([hashtable] $source, [psvariable] $dest) {
+    foreach ($key in $source.Keys) {
+        if ($dest.Value[$key]) {
```
Comment (weshaggard) [inline eng/common/TestResources/New-TestResources.ps1:113]: Instead of passing in the variable name why not pass in the dest as a hashtable as well? I would also suggest you use ".ContainsKey" here instead of indexing into the value directly.

File: eng/common/TestResources/New-TestResources.ps1 (hunk 17 above)
Comment (benbp) [reply eng/common/TestResources/New-TestResources.ps1:114]: How about I do a follow-on that changes them all? That way it's consistent across the script.

File: eng/common/TestResources/New-TestResources.ps1 (hunk 19 above)
Comment (benbp) [reply eng/common/TestResources/New-TestResources.ps1:113]: I wanted to be able to log the name of the hash so a log reader could understand which object had overrides against it (e.g. deployment outputs, template parameters, etc.). I think the two options for that are to pass a third variable representing the name, or do it this way.

File: eng/common/TestResources/New-TestResources.ps1 (hunk 18 above)
Comment (weshaggard) [reply eng/common/TestResources/New-TestResources.ps1:314]: If we end up with more cases maybe we split these into another env variable list and then use the same merge function you wrote for the others.

**Review by weshaggard — APPROVED** [review]

File: eng/common/TestResources/New-TestResources.ps1 (hunk 12 above)
Comment (heaths) [reply eng/common/TestResources/New-TestResources.ps1]: It still has to match the parameter type (if declared, like it is), so empty would be `@{}` which matches the default.

File: eng/common/TestResources/New-TestResources.ps1 (hunk 17 above)
Comment (heaths) [reply eng/common/TestResources/New-TestResources.ps1:114]: I recommend a `Warn` function like I did for logging that 1) always calls Write-Warning, and 2) if running under AzDO (or `$CI` is enabled, which would be consistent) uses the AzDO logging. Remember that we devs use this too and `##vsts.something` isn't really going to stick out.

Comment (benbp) [reply eng/common/TestResources/New-TestResources.ps1:114]: https://github.com/Azure/azure-sdk-tools/blob/master/eng/common/scripts/logging.ps1#L6

**Review by weshaggard — APPROVED** [review]
